RHPAM-2319: Stunner - Users don't understand how to create lines between nodes

romartin opened the pull request:
Hey @hasys @jomarko 

Here is the PR for https://issues.jboss.org/browse/RHPAM-2319

Notice I refactored most of the lienzo toolbox stuff, which was in `stunner-lienzo-extensions`, into the `lienzo-core`, so this PR includes the removal of several classes reated to toolbox, but those are included in `lienzo-core` / `lienzo-tests` now.

It depends on:
https://github.com/kiegroup/lienzo-core/pull/97
https://github.com/kiegroup/lienzo-tests/pull/78

Thanks!

romartin commented:
Jenkins execute full downstream build

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/test/java/org/kie/workbench/common/stunner/core/client/components/toolbox/actions/CreateNodeToolboxActionTest.java:
```diff
@@ -177,7 +177,10 @@ public void setup() throws Exception {
                                                   translationService,
-                                                  actions)
+                                                  actions,
+                                                  // TODO
```
TODO tests

romartin replied:
done

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/test/java/org/kie/workbench/common/stunner/core/client/components/toolbox/actions/FlowActionsToolboxFactoryTest.java:
```diff
@@ -203,10 +201,11 @@ public void testBuildToolbox() {
         verify(view,
                times(1)).init(eq(actionsToolbox));
-        verify(view,
+        // TODO: fix properly.
```
TODO tests

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/test/java/org/kie/workbench/common/stunner/core/client/components/toolbox/actions/MorphActionsToolboxFactoryTest.java:
```diff
@@ -176,10 +174,11 @@ public void testBuildToolbox() {
         verify(view,
                times(1)).init(eq(actionsToolbox));
-        verify(view,
+        // TODO: fix properly.
```
TODO tests

romartin replied:
done

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/command/DefaultSessionCommandManager.java:
```diff
@@ -18,150 +18,223 @@
 import org.kie.workbench.common.stunner.core.registry.command.CommandRegistry;
 
-import static org.kie.soup.commons.validation.PortablePreconditions.checkNotNull;
+// TODO: Is it really App scoped? test it while having multiple editors open...
```
TODO realization

romartin commented:
I know @jomarko @hasys 
Sorry for wasting your time, there are still a few issues on some last chagne I did. I'm trying to fix all those, will let you know ASAP. Thanks for the review and testing anyway, good to have that feedback now :+1+

romartin commented:
Hey @hasys @jomarko 

I've pushed some stuff, it makes it work better, but there are still a few issues and missing tests, which I'm working on... but DO NOT waste time yet revieweing it again plz, will let you know once ready, I'm close..

Thx in advance!

romartin commented:
Hey @hasys @jomarko 

I'm adding new commits just for not having all changes only in my local, but this is **still NOT ready** for review - I'm close, just missing to rebase, add a few tests and review dmn, anyway will let you know once ready.

Sorry for the inconveniences. Thx!

romartin commented:
Jenkins execute full downstream build

romartin commented:
@hasys @jomarko 

Hey guys, I just completed missing points and fixed some issues, also rebased from master, and this is now ready to review again! :)

So once you've a chance, plz go ahead and lemme know your feedback, thanks in advance!

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/components/toolbox/actions/CreateConnectorToolboxAction.java:
```diff
@@ -100,40 +79,41 @@ public String getEdgeId() {
+        connectorProxy.start(event);
     }
 
+    // TODO
```
done

romartin commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/canvas/controls/toolbox/DMNCommonActionsToolboxFactoryTest.java:
```diff
@@ -137,10 +134,11 @@ public void testBuildToolboxForNoDecisionType() {
         verify(viewInstance,
                times(1)).init(eq(actionsToolbox));
-        verify(viewInstance,
+        // TODO
```
done

manstis replied:
It looks like this `TODO` is however still in the PR.. should it be removed?

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/test/java/org/kie/workbench/common/stunner/core/client/components/toolbox/actions/CreateConnectorToolboxActionTest.java:
```diff
@@ -44,26 +35,19 @@
 import static org.mockito.Mockito.when;
 
-@RunWith(GwtMockitoTestRunner.class)
+// TODO: @RunWith(GwtMockitoTestRunner.class)
```
done

romartin commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/canvas/controls/toolbox/DMNCommonActionsToolboxFactoryTest.java:
```diff
@@ -205,10 +204,11 @@ public void testBuildToolboxForBusinessKnowledgeModelType() {
         verify(viewInstance,
                times(1)).init(eq(actionsToolbox));
-        verify(viewInstance,
+        // TODO
```
done

manstis replied:
It looks like this `TODO` is however still in the PR.. should it be removed?

romartin commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/canvas/controls/toolbox/DMNCommonActionsToolboxFactoryTest.java:
```diff
@@ -171,10 +169,11 @@ public void testBuildToolboxForDecisionType() {
         verify(viewInstance,
                times(1)).init(eq(actionsToolbox));
-        verify(viewInstance,
+        // TODO
```
done

manstis replied:
It looks like this `TODO` is however still in the PR.. should it be removed?

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/components/proxy/LienzoConnectorProxy.java:
```diff
@@ -0,0 +1,155 @@
+            return getConnectorView(connector);
+        }
+
+        // TODO
```
done

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/components/proxy/LienzoConnectorProxy.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    private void destroyProxy(WiresConnector connector) {
+        // TODO
```
done

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/components/proxy/LienzoNodeProxy.java:
```diff
@@ -0,0 +1,187 @@
+                commandFactory.addNode(getTargetNode(),
+                                       getShapeSetId());
+
+        // TODO: BPMN should provide an specific different magnet approach...
```
done

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/components/toolbox/actions/AbstractActionsToolboxView.java:
```diff
@@ -129,13 +110,77 @@ protected V init(final ActionsToolbox toolbox,
+        for (ToolboxAction toolboxAction : toolbox) {
+            final ButtonItem button = addButton(toolbox.getGlyph(toolboxAction),
+                                                toolbox.getTitle(toolboxAction));
+            // TODO final ItemsToolboxHighlight highlight = new ItemsToolboxHighlight(toolboxView);
```
done

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/components/proxy/LienzoNodeProxy.java:
```diff
@@ -0,0 +1,187 @@
+            return getTargetShapeView(targetShape);
+        }
+
+        // TODO
```
done

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/components/toolbox/actions/AbstractActionsToolboxView.java:
```diff
@@ -129,13 +110,77 @@ protected V init(final ActionsToolbox toolbox,
+
+                button
+                        .onMoveStart(event -> {
+                            // TODO highlight.highlight(button);
```
done

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/test/java/org/kie/workbench/common/stunner/client/lienzo/components/toolbox/actions/CommonActionsToolboxViewTest.java:
```diff
@@ -36,7 +34,8 @@
 import static org.mockito.Mockito.when;
 
-@RunWith(LienzoMockitoTestRunner.class)
+// TODO: @RunWith(LienzoMockitoTestRunner.class)
```
done

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/test/java/org/kie/workbench/common/stunner/core/client/components/toolbox/actions/ActionsToolboxTest.java:
```diff
@@ -112,14 +109,15 @@ public void testInit() {
         verify(view,
                times(1)).init(eq(tested));
-        verify(view,
+        // TODO: fix properly.
```
done

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/test/java/org/kie/workbench/common/stunner/core/client/components/toolbox/actions/CommonActionsToolboxFactoryTest.java:
```diff
@@ -110,10 +106,11 @@ public void testBuildToolbox() {
         verify(view,
                times(1)).init(eq(actionsToolbox));
-        verify(view,
+        // TODO: fix properly.
```
done

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/test/java/org/kie/workbench/common/stunner/client/lienzo/components/toolbox/actions/FlowActionsToolboxViewTest.java:
```diff
@@ -36,7 +34,8 @@
 import static org.mockito.Mockito.when;
 
-@RunWith(LienzoMockitoTestRunner.class)
+// TODO: @RunWith(LienzoMockitoTestRunner.class)
```
done

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/components/toolbox/actions/AbstractActionsToolboxView.java:
```diff
@@ -129,13 +110,77 @@ protected V init(final ActionsToolbox toolbox,
+                // Highlight the toolbox button once cliking on it. No need to call restore after this,
+                // because Stunner actually destroys and recreates the toolbox after click
+                // (so after some command has been executed).
+                // TODO highlight.highlight(button);
```
done

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/components/proxy/LienzoNodeProxy.java:
```diff
@@ -0,0 +1,187 @@
+    }
+
+    private void destroyProxy(final WiresShape shape) {
+        // TODO
```
done

romartin commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/canvas/controls/toolbox/DMNFlowActionsToolboxFactoryTest.java:
```diff
@@ -168,9 +165,10 @@ public void testBuildToolbox() {
         verify(viewInstance,
                times(1)).init(eq(actionsToolbox));
-        verify(viewInstance,
+        // TODO
```
done

manstis replied:
It looks like this `TODO` is however still in the PR.. should it be removed?

romartin commented:
Jenkins execute full downstream build

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/components/toolbox/actions/FlowActionsToolboxView.java:
```diff
@@ -13,20 +13,31 @@
  * limitations under the License.
  */
-
+// TODO
```
I saw more but didn't check it properly, will do it on Monday.

manstis replied:
It looks like this `TODO` is however still in the PR.. should it be removed?

hasys commented:
Hi @lclay2,

you can download build and test it on your. Now we have a nice feature in Full Downstream Build and it is really easy to try locally. There are steps how to do that:
1. Open `Linux - full downstream` by clicking on `Details` link
![image](https://user-images.githubusercontent.com/1477262/70907469-3a1d1380-2009-11ea-950f-cb0621e3efc5.png)
2. Download `jbpm-server-7.31.0-SNAPSHOT-dist.zip` file and unzip it on your machine 
![image](https://user-images.githubusercontent.com/1477262/70907489-499c5c80-2009-11ea-9893-8e9a509557ba.png)
3. Go to `jbpm-server-7.32.0-SNAPSHOT-dist/bin` folder in the terminal and execute `./standalone.sh` and give 30/60 seconds to the server to start up.
4. Open in you browser `http://localhost:8080/business-central` and login with login `krisv` and password `krisv`
5. Create new project and new process and you can test it locally on your machine :)

lclay2 commented:
@hasys  thanks for the instructions, I will try.

hasys commented:
> @hasys thanks for the instructions, I will try.

Forgot to say that you need a Java 8+ on your computer to be able to start up the server. But with this new Jenkins feature with full package it will be possible to test all PRs easily (except for Kogito for now, but it should be available soon as well). So it make sense to me to install Java 8 once and benefits from it for every PRs in the future :)

lclay2 commented:
@hasys  @romartin  It worked, I have it running locally now. Thanks so much! :) @bdellasc and I will review it together tomorrow.

romartin commented:
Great, thanks  @lclay2 @bdellasc !!

hasys commented:
Hi @romartin,

code looks good for all 3 parts (kie-wb-common, lienzo-core, lienzo-tests) but missing some coverage for conditions in kie-wb-common, as we discussed it is not related to code moving it is new changes mostly, because Sonar doesn't count lienzo-core coverage for this repository.

I guess all works good, since @lclay2 and @bdellasc already tested it at runtime, but I will check it for different browsers as well.

Thank you @romartin :) we are very close to finish this PR!

hasys commented:
Jenkins execute full downstream build

(if build fails, it can be caused by https://github.com/kiegroup/kie-wb-common/pull/3095 but, maybe it works since branch was created some time ago and possibly not affected).

romartin commented:
Jenkins retest this

romartin commented:
Jenkins exectue full downstream build

romartin commented:
Jenkins execute full downstream build

romartin commented:
hey @manstis 

good point, I can do the same in BPMN, I mean create a connection to nowhere.... it should be not possible yeah, will fix it, working on it.

Also I'll try to reproduce the bug you mention, I was not aware of it, thanks for helping @manstis!! Will let you know once both fixed so you can give it a try again... :)

romartin commented:
Dunno why builds are failing, cannot reproduce locally.... rebasing and rebuilding

romartin commented:
Jenkins retest this

romartin commented:
Jenkins exectue full downstream build

domhanak commented:
What's wrong with FDB? :)

romartin commented:
Jenkins retest his

romartin commented:
Jenkins execute full downstream build

jomarko commented:
@manstis can you help with failing test?
```
    Invoking generator org.jboss.errai.ioc.rebind.ioc.bootstrapper.IOCGenerator
            [ERROR] Error generating BootstrapperImpl
java.util.concurrent.ExecutionException: java.lang.RuntimeException: The following dependency problems were found:
	Unsatisfied constructor dependency @Default org.kie.workbench.common.stunner.core.client.api.SessionManager for Class @Any @Default org.kie.workbench.common.stunner.core.client.i18n.ClientTranslationService. Some beans were found that satisfied this dependency, but must be enabled:
org.kie.workbench.common.stunner.core.client.api.GlobalSessionManager


	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at org.jboss.errai.config.rebind.AsyncGenerators$FutureWrapper.get(AsyncGenerators.java:112)
	at org.jboss.errai.config.rebind.AsyncGenerators$FutureWrapper.get(AsyncGenerators.java:86)
	at org.jboss.errai.config.rebind.AbstractAsyncGenerator.startAsyncGeneratorsAndWaitFor(AbstractAsyncGenerator.java:205)
	at org.jboss.errai.ioc.rebind.ioc.bootstrapper.IOCGenerator.generate(IOCGenerator.java:62)
	at com.google.gwt.core.ext.IncrementalGenerator.generateNonIncrementally(IncrementalGenerator.java:40)
	at com.google.gwt.dev.javac.StandardGeneratorContext.runGeneratorIncrementally(StandardGeneratorContext.java:745)
	at com.google.gwt.dev.cfg.RuleGenerateWith.realize(RuleGenerateWith.java:103)
	at com.google.gwt.dev.shell.StandardRebindOracle$Rebinder.rebind(StandardRebindOracle.java:78)
	at com.google.gwt.dev.shell.StandardRebindOracle.rebind(StandardRebindOracle.java:262)
	at com.google.gwt.dev.shell.StandardRebindOracle.rebind(StandardRebindOracle.java:251)
	at com.google.gwt.dev.PrecompilationContextCreator$1.getAllPossibleRebindAnswers(PrecompilationContextCreator.java:86)
	at com.google.gwt.dev.jjs.impl.UnifyAst$UnifyVisitor.createStaticRebindExpression(UnifyAst.java:519)
	at com.google.gwt.dev.jjs.impl.UnifyAst$UnifyVisitor.createRebindExpression(UnifyAst.java:487)
	at com.google.gwt.dev.jjs.impl.UnifyAst$UnifyVisitor.maybeHandleMagicMethodCall(UnifyAst.java:415)
	at com.google.gwt.dev.jjs.impl.UnifyAst$UnifyVisitor.visit(UnifyAst.java:402)
	at com.google.gwt.dev.jjs.ast.JMethodCall.traverse(JMethodCall.java:265)
	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:265)
	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:118)
	at com.google.gwt.dev.jjs.ast.JCastOperation.traverse(JCastOperation.java:76)
	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:265)
	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:118)
	at com.google.gwt.dev.jjs.ast.JDeclarationStatement.traverse(JDeclarationStatement.java:49)
	at com.google.gwt.dev.jjs.ast.JModVisitor$ListContext.traverse(JModVisitor.java:88)
	at com.google.gwt.dev.jjs.ast.JModVisitor.acceptWithInsertRemove(JModVisitor.java:331)
	at com.google.gwt.dev.jjs.ast.JBlock.traverse(JBlock.java:94)
	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:139)
	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:135)
	at com.google.gwt.dev.jjs.ast.JTryStatement.traverse(JTryStatement.java:93)
	at com.google.gwt.dev.jjs.ast.JModVisitor$ListContext.traverse(JModVisitor.java:88)
	at com.google.gwt.dev.jjs.ast.JModVisitor.acceptWithInsertRemove(JModVisitor.java:331)
	at com.google.gwt.dev.jjs.ast.JBlock.traverse(JBlock.java:94)
	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:139)
	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:135)
	at com.google.gwt.dev.jjs.ast.JMethodBody.traverse(JMethodBody.java:83)
	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:265)
	at com.google.gwt.dev.jjs.ast.JMethod.visitChildren(JMethod.java:786)
	at com.google.gwt.dev.jjs.ast.JMethod.traverse(JMethod.java:778)
	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:265)
	at com.google.gwt.dev.jjs.impl.UnifyAst.mainLoop(UnifyAst.java:1401)
	at com.google.gwt.dev.jjs.impl.UnifyAst.exec(UnifyAst.java:896)
	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.unifyJavaAst(JavaToJavaScriptCompiler.java:1410)
	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.constructJavaAst(JavaToJavaScriptCompiler.java:1222)
	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.precompile(JavaToJavaScriptCompiler.java:1140)
	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.precompile(JavaToJavaScriptCompiler.java:255)
	at com.google.gwt.dev.Precompile.precompile(Precompile.java:255)
	at com.google.gwt.dev.Precompile.precompile(Precompile.java:202)
	at com.google.gwt.dev.Precompile.precompile(Precompile.java:143)
	at com.google.gwt.dev.Compiler.compile(Compiler.java:204)
	at com.google.gwt.dev.Compiler.compile(Compiler.java:155)
	at com.google.gwt.junit.JUnitShell.compileForWebMode(JUnitShell.java:1102)
	at com.google.gwt.junit.JUnitShell.maybeCompileForWebMode(JUnitShell.java:1167)
	at com.google.gwt.junit.CompileStrategy.maybeCompileModuleImpl2(CompileStrategy.java:183)
	at com.google.gwt.junit.CompileStrategy.maybeCompileModuleImpl(CompileStrategy.java:113)
	at com.google.gwt.junit.SimpleCompileStrategy.maybeCompileModule(SimpleCompileStrategy.java:36)
	at com.google.gwt.junit.JUnitShell.runTestImpl(JUnitShell.java:1358)
	at com.google.gwt.junit.JUnitShell.runTestImpl(JUnitShell.java:1326)
	at com.google.gwt.junit.JUnitShell.runTest(JUnitShell.java:682)
	at com.google.gwt.junit.client.GWTTestCase.runTest(GWTTestCase.java:421)
	at junit.framework.TestCase.runBare(TestCase.java:141)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:129)
	at com.google.gwt.junit.client.GWTTestCase.run(GWTTestCase.java:247)
	at junit.framework.TestSuite.runTest(TestSuite.java:252)
	at junit.framework.TestSuite.run(TestSuite.java:247)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:86)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: java.lang.RuntimeException: The following dependency problems were found:
	Unsatisfied constructor dependency @Default org.kie.workbench.common.stunner.core.client.api.SessionManager for Class @Any @Default org.kie.workbench.common.stunner.core.client.i18n.ClientTranslationService. Some beans were found that satisfied this dependency, but must be enabled:
org.kie.workbench.common.stunner.core.client.api.GlobalSessionManager
```

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-client-api/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/Canvas.java:
```diff
@@ -66,6 +66,7 @@
     Canvas clear();
 
     /**
+     * TODO: Remove - never called?
```
It looks like this `TODO` is still in the PR.. should it be removed?

manstis commented:
@jomarko @romartin I looked at the failing test however there's nothing in that DMN module to directly lead to the error:-
```
Unsatisfied constructor dependency @Default org.kie.workbench.common.stunner.core.client.api.SessionManager for Class @Any @Default org.kie.workbench.common.stunner.core.client.i18n.ClientTranslationService. Some beans were found that satisfied this dependency, but must be enabled:
org.kie.workbench.common.stunner.core.client.api.GlobalSessionManager
```
I see this PR changes the injection of `SessionManager` so perhaps there's an occurrence/change missing?

romartin commented:
@manstis @jomarko no worries guys, I'll take a look at it, I did some changes that may affect yes.... 

Thanks for helpping, but do not spend time plz, I'm feeling guilty... :) just lemme know issues or whatever is missing and I'll do that, thanks anyway for all your help! 

I'm still doing some last cleans here, will let you know once ready to review again, also waiting for Jenkins to work again...

Thx!

romartin commented:
@manstis @jomarko 

Found the fix on the DMN test, see https://github.com/kiegroup/kie-wb-common/pull/3012/commits/ef3b7f65c1f8de8be11e2781acb255dd2b485ef1#diff-fdc336af881bce173ca25cb29c242e48R52

I just renamed the class... can those classnames be, in someway, refactored and removed from the `ErraiApp`? Just asking, not sure if it's possible, but would help on further refactors/changes...

Thx guys!

romartin commented:
Jenkins retest this

romartin commented:
Jenkins retest this

romartin commented:
Jenkins execute full downstream build

manstis commented:
@romartin 

> can those classnames be, in someway, refactored and removed from the `ErraiApp`?

Unfortunately not easily. The unit tests in `kie-wb-common-dmn-webapp-kogito-marshalling-tests` inherit an Errai test class in an Errai module that contains various _whitelist_ and _blacklist_ files (that drive what classes Errai's CDI scan). This led to me needing to declare the Stunner/DMN classes in a `errai.ioc.whitelist` in order for them to be discovered by Errai CDI to ensure constructor injection points were satisfied.

romartin commented:
Jenkins execute full downstream build

romartin commented:
Jenkins execute full downstream build

romartin commented:
Jenkins execute full downstream build

romartin commented:
Jenkins execute full downstream build

romartin commented:
Hey @domhanak @manstis ( @jomarko )

I just rebased this OLD PR again and re-fired the build jobs. Anyway, as we're still running into issues with the FDB, I've uploaded the (locally) generated [Business Central](https://drive.google.com/file/d/1bFj2IiiFgVJQqLVUueV5cFsWTAcw1Y_B/view?usp=sharing) (WAR) so you can test the UI.

This is an **urgent PR for 7.7**, plz can you provide some feedback on the review again? Really appreciated!!

jomarko commented:
@romartin, FYI @domhanak is on PTO rest of the week.

romartin commented:
Jenkins retest this

romartin commented:
Jenkins execute full downstream build

romartin commented:
@hasys @manstis @domhanak your review is missing please. thx in advance! :+1: 

BTW Removing the uploaded WAR as FDB seem to be working...

romartin commented:
Hey @LuboTerifaj - just noticed @domhanak is on PTO, but his review is still missing. Can you handle that? This is URGENT for a release, please. Thx in advance!

mbiarnes commented:
Jenkins please execute full downstream build

domhanak commented:
Hi @romartin, works well. However I run into some issues while testing acceptance criteria in both Chrome and Firefox.

Issues no.1 **Critical/Blocker**
You can model multiple sequence flow out of start event and the process is still validated.
Example:
![Screenshot from 2020-02-03 08-43-16](https://user-images.githubusercontent.com/5937251/73634333-487ac980-4661-11ea-8f2f-42fd1239c22b.png)

_________________________________________________________________________

Issues no.2 **Critical**
When you want to Click&Drag using quick menu in embedded sub process an Unexpected Error occurs. Just create a embedded sub-process and try to model in it using Drag option, not just clicks.
Notification error:
```
Command execution failed.
Reason/s:
[1] Sequence flow connectors cannot exceed the parent's subprocess bounds. Both source and target nodes must be in the same subprocess.!org.kie.workbench.common.stunner.core.rule.violations.RuleViolationImpl!!!
```

Error in console:
```
08:41:06,446 ERROR [org.kie.workbench.common.services.backend.logger.GenericErrorLoggerServiceImpl] (default task-14) Error from user: krisv Error ID: -1047165319 Location: LibraryPerspective|$ProjectScreen,org.kie.workbench.common.screens.messageconsole.MessageConsole,AddAssetsScreen[EDiagramEditorPropertiesScreen,],,BPMNDiagramEditor?path_uri=default://master@MySpace/RHPAM-2319/src/main/resources/com/myspace/rhpam_2319/Process.bpmn&file_name=Process.bpmn&has_version_support=true Exception: Uncaught exception: Exception caught: Exception caught: (TypeError) : e is undefined Caused by: Exception caught: (TypeError) : e is undefined Caused by: (TypeError) : e is undefined
```
Workaround is to just click.

______________________________________________________________________________

Issues no.3 **Critical/Major** 
This shouldn't be allowed as it guides user towards incorrect processes.

It is possible to connect a sequence flow from Activity to itself. Creating this: 
![Screenshot from 2020-02-03 08-24-05](https://user-images.githubusercontent.com/5937251/73633292-96420280-465e-11ea-83b2-b2b522bc5acb.png)
It is only possible when I use quick menu (drag or click) to create sequence flow. This should not be allowed. 
Steps to reproduce:
1.  Create an Activity (task, sub-process,..)
2.  Use quick menu to add sequence flow (draog or click, both work)
3.  Click on top of the source Activity.

In case of Events, the Node get highlighted in red and no connection is made. This should be same for activities as this behavior should be implemented using gateways.

domhanak commented:
@romartin in my comment I highlighted some issues, that are also possible in previous versions. So lets have a call when you can to process and see what really is connected with RHPAM-2319 :)  It looked to me that they are connected to this PR, but I am able to reproduce them with older builds:
Like start event having multiple outgoing connections (bad practice, but we allow doing it.)

romartin commented:
Hey @krisv - quick Q plz about the [issue #3](https://github.com/kiegroup/kie-wb-common/pull/3012#issuecomment-581281309) reported by @domhanak 

As far as I remember, it should possible to model a sequence flow, for an activity, that targets itself (eg: loops), right? (I remember fixing an issue related to this time ago...)

On the other hand, when modelling it, although it's actually allowed, it looks pretty bad (see above screenshot)... I cannot even create new bendpoints, but nobody complained about it... you know how people uses it?

krisv commented:
@romartin I would argue that tasks looping too themselves doesn't really make sense.  Maybe if we were supporting multiple connections (which is not available in stunner and as discussed before we strongly recommend users to use explicit gateways) but even then it would be an edge case imho.  I don't see improving this as a priority.

romartin commented:
ok thanks @krisv , but should we also PREVENT for creating those (form the UI)? (We actually ALLOW to do that)

krisv commented:
@romartin not sure it's necessary to prevent it, we could consider it but in the end I don't think it's explicitly forbidden to do it either.  So I wouldn't see preventing it as a priority either ;)

romartin commented:
ok thanks for the feedback @krisv 

So doing a re-cap again @domhanak - only missing `issue #2`, trying to reproduce it now... 

thanks!

romartin commented:
Hey @domhanak 

I'd been spending a few hours on the issue `#2`, great catch!! I'm close to have the right fix but I'm still running into some issues, so not yet feeling confident for adding a new commit that fixes it. 

On the other hand, this PR is big, old and it contains several other refactorings. I would really like to introduce this new code ASAP, trying to get earlier feedback from others and avoiding the maintaince of this PR. 

Soo what do you think about moving forward as:

1. Reporting the issues `#2`  and `#3` (as  `#1` has already [been reported](https://issues.redhat.com/browse/RHPAM-2671))

2. Link the issues to this [RHPAM-2319](https://issues.redhat.com/browse/RHPAM-2319)

3. Mark the issue `#2` as a blocker for [RHPAM-2319](https://issues.redhat.com/browse/RHPAM-2319) - So I'll continue working on it in a new PR / commit during tomorrow

This way, you could approve, I could merge this big work and keep working at same time on the fix for subprocesses. I would really help to save work, at least from my side. WDYT??? 

PS: SonarCloud checks are mostly complaining about refactored code, which previous to this work it has not the right tests before, so same status now.... Given the urgency of this stuff, I think we should not block the feature due to it. Anyway, also feel free to provide your feedback and opinion!

Thanks!

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/palette/BS3PaletteWidgetViewImpl.java:
```diff
@@ -71,49 +65,49 @@ public void showDragProxy(String itemId,
+                                       @Override
+                                       public void onComplete(int x,
+                                                              int y) {
+                                           presenter.onDragProxyComplete(itemId,
```
as commented with Kris below, that's not necessary a bug. We can create a new ticket for preventing it,from the UI (Stunner), although not high priority. Thanks anyway for pointing out, good catch!

romartin commented:
Jenkins retest this

romartin commented:
Jenkins execute full downstream build

romartin commented:
Hey @domhanak 

I've been able to fix the issue `#2`! Thanks for reporting!! 

So please, see my comment above, let's consider reporting issues `#1` and `#3` and try to merge this ASAP.

@domhanak **URGENT** Can you please complete your review? Thx in advance!

domhanak commented:
jenkins retest this

domhanak commented:
@krisv @romartin Keep in mind that if user connects activity to itself he is not allowed to add additional sequence flow,  as we allow only one seq flow from activity.

domhanak commented:
@manstis we still miss your approval

manstis commented:
@domhanak I still see commented out "TODO's" in the DMN changes.

Nobody has advised whether they're to be removed, fixed or the subject of another JIRA.

romartin commented:
@manstis hey! I'm just trying to look for, but which TODs have been commented out? I cannot find any in DMN, can you please point me to those? Also, it's missing your approval here yet, I was going to merge this.... can you please give it a try? (anywayt let you know I also tried it several times anyway, I'm always testing both editors once adding some new commit)
Thanks in advance!!

romartin commented:
Jenkins retest this

romartin commented:
Jenkins execute full downstream build

romartin commented:
Jenkins retest this

romartin commented:
@manstis sorry for those old TODOs, my fault!! Thanks for letting me know, all ready again! :+1:

romartin commented:
Jenkins retest this

romartin commented:
Jenkins execute full downstream build

romartin commented:
The [regular buid](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/KIE/job/master/job/pullrequest/job/kie-wb-common-pullrequests/1540/) is continuously failing due to this "unrelated" test case: `org.guvnor.ala.services.rest.tests.RestRuntimeProvisioningImplTest.testAPI`

The [full build](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/KIE/job/master/job/pullrequest/job/kie-wb-common-downstream-pullrequests/909/) has been running with no issues.

Also tested both BPMN/DMN editors on the generated artifacts and everything works properly.

So merging! Thanks everyone for the good work!
